### ISSUE TITLE
Support for Cyrlic names in whisper window for character names

### ISSUE THREAD

**Issue by PedjaSenpai** [Legacy-of-Sylvanaar/wow-instant-messenger#80]
Is it possible to add option to set font, or where to change font that default skin is using to show whispering character names if its in Cyrlics, it shows color of class correct but name is [][][][][][][][][][][]

**Comment by Pazza** [comment]
You could create your own skin using only the skin.lua and skin.toc file found in WIM/Skins/Default. You would piece it together as a new addon. You only need to include the actual keys you want changed.

It is using FriendsFont_Normal which you would think is localized for your client.

**Comment by Antrakt92** [comment]
I am seeing the same problem with WIM 3.17.3 on Retail when the sender has a Cyrillic name. The message body can render correctly after choosing a Cyrillic-capable font in **/wim > Display Settings**, but the sender name in the window title does not. Cyrillic names in the conversation picker opened from the WIM minimap icon can fail for the same reason.

I traced the current code and the selected WIM font does not reach either of those labels:

- `Skins/Default/skin.lua` hard-codes `message_window.widgets.from.font = "FriendsFont_Normal"`.
- `Sources/Skinner.lua` applies `db.skin.font` only to `chat_display` and `msg_box`.
- `Modules/Menu.lua` uses the `UIPanelButtonTemplate` label (`button.text`) for conversation names and never applies the selected WIM font to it.

This matters because the title and minimap conversation picker are the main ways to identify who sent the whisper and which conversation to open.

My update-safe workaround is a tiny separate addon loaded alongside WIM (it does not edit WIM files). It:

1. calls `SetFont("Fonts\\ARIALN.TTF", existingSize, existingFlags)` on each window's `widgets.from`;
2. applies the same font to `WIM.Menu.groups[*].buttons[*].text`;
3. hooks `WIM.ApplySkinToWindow` and reapplies the title font after WIM changes/reapplies a skin.

The built-in Arial Narrow font renders Cyrillic correctly on my client, and keeping this in a separate addon means WIM updates do not overwrite the workaround.

For an upstream fix, could the title and menu labels follow the user's selected `db.skin.font`, with their existing size/flags preserved? A locale-capable fallback such as guarded `STANDARD_TEXT_FONT` / Arial Narrow would also work. A custom skin can fix the title, as suggested above, but it does not appear to cover the separate labels created in `Modules/Menu.lua`, and locale support ideally should not require users to maintain a second addon.

**Comment by Pazza** [comment]
Thank you for taking the time to provide all that insight. When I originally developed it I assumed that these fonts would support whatever fonts it's game client would cater to... I guess I was wrong. I think the normal text don't would probably be the easiest path to take. I will take a look at it. I'm surprised chat font normal wouldnfailmto support it to be honest since the chat frame sees it all... 🤷

**Comment by Pazza** [comment]
@Antrakt92 I've been poking around here. Seems like FriendsFont_Normal is the same as STANDARD_TEXT_FONT and on the western game client's they don't include the cryllic characters, so another font would have to be used. I don't want to change the way the default skin looks... Other thoughts? I could include a cryllic skin which uses the fonts that would make it work... 

Regarding the menu, I'm thinking of the best path to apply the skin there as well. I might just inherit the from font of the selected skin.

Sorry, one more thing, I don't see where widget.from.font is hardcoded.

**Comment by Pazza** [comment]
Do you think it would be OK to look at localization and programmatically change the default skin font based on that?

for example, for:
ruRU: STANDARD_TEXT_FONT

and whichever other locale should use the same definition. Or is the problem you are using a western game client?

**Comment by Antrakt92** [comment]
@Pazza Yes — this is the Western-client case. My Retail client is configured with `textLocale "enUS"` and `audioLocale "enUS"`, on the EU portal. The sender had a Cyrillic character name, so this is Cyrillic text arriving on an English client.

Thanks for looking into this so quickly. I pulled the current `master` and saw the new [`OnSkinLoaded` event](https://github.com/Legacy-of-Sylvanaar/wow-instant-messenger/commit/e52f14e4713c3f62cae3985844da72c0befbefb6) and the [menu/user-list skinning work](https://github.com/Legacy-of-Sylvanaar/wow-instant-messenger/commit/5cb98b2df28f9c6420feb87fa8bb265bfae21dfc). That fixes the limitation I mentioned earlier where the conversation menu was outside the skin system.

Also, sorry that my “hard-coded” shorthand was unclear. I meant this nested entry inside `WIM_ClassicSkin`, not a literal assignment named `widget.from.font`:

```lua
message_window = {
    widgets = {
        from = {
            font = "FriendsFont_Normal",
```

It is [here in the current file](https://github.com/Legacy-of-Sylvanaar/wow-instant-messenger/blob/3a21783380c6bbff7583321c8bfa48f43a5b1f7d/Skins/Default/skin.lua#L118-L126).

I think locale-based selection can still be useful for native locale defaults, but it would not cover this particular case. A `ruRU` branch would never run on this `enUS` client, even though the displayed name is Cyrillic. As you found, `FriendsFont_Normal` and `STANDARD_TEXT_FONT` both resolve to the Western FRIZ font here, so changing only the `ruRU` default would leave this reproduction unchanged.

A separate Cyrillic skin could work as an optional preset, but it ties glyph coverage to the whole visual theme. Someone using Classic Blue or a third-party skin would have to give that up, or every skin would need a Cyrillic variant.

While checking the current changes I noticed a few implementation details that may be worth handling before release:

- `menu.button.font` is currently a separate field set to `FriendsFont_Normal`; it does not actually inherit `message_window.widgets.from.font`. A custom skin that changes only the sender-name font would therefore still leave the conversation menu unchanged.
- The new menu code assumes the value is the name of a global FontObject and calls `_G[fontName]:GetFont()`. The WIM font picker normally stores an LSM key such as `Arial Narrow`, while `STANDARD_TEXT_FONT` itself is a file-path string. Passing either directly through the current menu lookup can fail. Reusing or extending the resolver already used by `SetWidgetFont` for both FontObjects and LSM fonts seems safer.
- `OnSkinLoaded` changes the menu font but does not call `Refresh()`. Menu width is measured from `GetStringWidth()` during refresh, so a wider font can leave the old width until another window event happens.
- Since the old hard-coded menu backdrop was moved into `ApplySkin`, there is also a fallback edge if `db.skin.selected` names a custom skin that was removed or did not load. In that case no `LoadSkin` event is fired for the saved name, and `Menu:OnEnable` currently creates the menu without applying `GetSelectedSkin()`. Applying the fallback skin when the menu is created would make that path safe.

The smallest behavior change that seems consistent with the existing settings would be:

- When **Use font suggested by skin** is enabled, keep the current skin appearance and let the conversation-name labels use the skin’s effective `from` font face.
- When it is disabled, resolve the user’s existing `db.skin.font` selection and apply that face to the sender title and conversation/user-list labels as well as the message area, while preserving each label’s own size, flags and color.
- Reapply it after skin changes and refresh the menu layout afterwards.

That would solve this `enUS`/Cyrillic case without changing the default WIM Classic look for everyone.

One other related surface is tabbed windows: tab labels display `theUser` through `ChatFontNormal`, and the existing `tab_strip.text` font settings do not appear to be applied. The History user list is similar. They could be follow-up scope, but otherwise the name may be fixed in the title and minimap menu while still appearing as boxes in tabs/history.

I would be happy to test an unreleased build with a real Cyrillic sender name, font/skin toggles and `/reload`.

**Comment by Pazza** [comment]
- You are correct about `menu.button.font` not inherriting, it has been given its own font selection. I did this to add a little more flexibility to skinning the menus, otherwise, it would have been a lot simpler to just use the `widget.from.font` value. But I figured why not open it up more.
- I decided to used objects such as ChatFontNormal because they are defined by the client instead of using a defined font string which might change from client to client. The option provided to define text isn't handled a bit differently using SharedMediaLib and therefore is saved differently. I see your point though and I did consider writing in support to take both an object and a string. I might still do this. (but it doesn't fix our main problem yet).
- Good catch on calling `Menu:Refresh()` on font change.
- Regarding `LoadSkin` not being called if a skin is missing... This is true, although extremely rare. Truthfully its only an issue for third party addons providing skins. I'd have to see if there were a trivial solution to make sure that `OnSkinLoaded` gets called if the selected skin is not loaded yet... I suppose a time our need to be set to allow a grace period for it to load if not loaded yet.

All in all the best solution might be as stated and to provide another option essentially allowing the user to override fonts where names are used. OR, i keep it simple and use the the existing option to apply the supplied font to other areas where names are output... As I write it, this might makes the most sense, because this really is an edge case working outside of how the game client was designed.

**Comment by Pazza** [comment]
@Antrakt92 Give the latest alpha a go. You can see the commits I made. I think I covered all of our issues. (I didn't look at history viewer yet, but I think that should be using ChatFontNormal which I would think covers everything.

**Comment by Antrakt92** [comment]
@Pazza I tested `3.17.3-7-g6398a4f` on live Retail 12.1.0 after completely removing my DimaQoL WIM font workaround. The new code works: with Arial Narrow selected and **Use font suggested by skin** turned off, the same Cyrillic sender name renders correctly in the window title. Thank you — this solves my case.

The remaining concern is discoverability. In the first screenshot Arial Narrow is visibly selected, but because **Use font suggested by skin** is still checked, WIM continues using the skin font and the name remains boxes. A user seeing this problem is very unlikely to guess that they must both select a capable font and disable a separate option.

Perhaps selecting a font could automatically turn off **Use font suggested by skin**, or the font list could be visibly disabled/dimmed while the skin option is active. Ideally an automatic Cyrillic-capable fallback for conversation names containing Cyrillic would be even friendlier. Client locale alone cannot cover this case because this is an `enUS` client receiving a Cyrillic name.

**Before:** Arial Narrow selected, but the skin font is still active.

<img width="723" height="402" alt="Before: Arial Narrow selected while the skin font is still active" src="https://github.com/user-attachments/assets/a7b5a467-8eb2-4f0e-b7d6-66d8c36505e4" />

**After:** the skin-font option is disabled and the same name renders correctly.

<img width="712" height="423" alt="After: the skin font is disabled and the Cyrillic name renders correctly" src="https://github.com/user-attachments/assets/f001b5fe-3522-460b-ac06-d2e486f7eb69" />

I have confirmed the window title so far; I have not tested the History viewer yet.

**Comment by Pazza** [comment]
I agree with your concerns regarding the options and it not being very intuitive. Options might be rewritten, so until then... I think we're good to go for now.
